lib/rpmfi.c

```diff
@@ -1507,15 +1509,38 @@ static void rpmfilesBuildNLink(rpmfiles fi, Header h)
 }
 
 /* Convert a tag of hex strings to binary presentation */
-static uint8_t *hex2bin(Header h, rpmTagVal tag, rpm_count_t num, size_t len)
+/* If lengths is non-NULL, assume variable length strings */
+static uint8_t *hex2bin(Header h, rpmTagVal tag, rpm_count_t num, size_t len,
+			int **lengths, int *maxlen)
 {
     struct rpmtd_s td;
     uint8_t *bin = NULL;
 
     if (headerGet(h, tag, &td, HEADERGET_MINMEM) && rpmtdCount(&td) == num) {
-	uint8_t *t = bin = xmalloc(num * len);
 	const char *s;
 
+	/* Figure string sizes + max length for allocation purposes */
+	if (lengths) {
+	    int maxl = 0;
+	    int *lens = xmalloc(num * sizeof(*lens));
+	    int i = 0;
+
+	    while ((s = rpmtdNextString(&td))) {
+		lens[i] = strlen(s) / 2;
```

Comment:
What if `strlen(s)` is odd?  That should be an error.

Reply:
I guess the assumption here is that a hex string will always be even (one byte is represented by two chars), but otherwise yeah, we should probably have a check here.

Reply:
Right, these tags should be sanity checked along everything else in the signature header, but that doesn't currently happen for data that is not used for content checking. The same goes for fsverity signatures. Not that it matters a whole lot, the signature will be invalid in that case anyhow.